backend/backend.py

```diff
@@ -0,0 +1,157 @@
+from abc import ABC, abstractmethod
+from json import dumps
+from pathlib import Path
+from shutil import disk_usage
+from time import time
+from traceback import format_exc
+from typing import Callable
+
+from canvas import Canvas
+
+# A function which accepts a string and returns None
+PublishFunc = Callable[[str], None]
+
+class Backend(ABC):
+    """ Backend for getting and processing video feed.
+
+        Handles combining the video feed with overlays and displaying, and
+        recording the video feed to a file. """
+
+    def __init__(self, width: int, height: int, publish_recording_status_func: PublishFunc):
+        self.width = width
+        self.height = height
+        self.publish_recording_status_func = publish_recording_status_func
+
+        self.recording = False
+        self.recording_output_file = None
+        self.recording_start_time = None
+
+        # time that we last called self.send_recording_status
+        self.prev_recording_status_time = 0
+        # Time between recording statuses, in seconds
+        self.recording_status_interval = 60
+
+    @abstractmethod
+    def start_video(self) -> None:
+        """ Starts the necessary processes to begin displaying camera feed
+            video. """
+
+    @abstractmethod
+    def on_base_canvas_updated(self, base_canvas: Canvas) -> None:
+        """ Updates the base canvas which is drawn on the video. Should be
+            called whenever the canvas is updated. """
+
+    @abstractmethod
+    def on_canvases_updated(self, data_canvas: Canvas, message_canvas: Canvas) -> None:
+        """ Updates the data and message canvases which are drawn on the
+            video. Should be called whenever the canvases are updated. """
+
+    def on_loop(self) -> None:
+        """ Executes all operations that should run every iteration of the
+            Overlay loop.
+
+            This includes sending the recording status (if it is due) and
+            depending on the backend, the display may be updated during this
+            call. This operation may be blocking to ensure the display is
+            updated at the correct framerate. """
+
+        self._on_loop()
+
+        if time() > self.prev_recording_status_time + self.recording_status_interval:
+            self.send_recording_status()
+
+    @abstractmethod
+    def _on_loop(self) -> None:
```

Comment:
Derived classes override `_on_loop` here but not `on_loop`, but in other places where there aren't two methods, derived classes would override a method without an underscore. Is this fine, or is there a convention that makes more sense I should use?

Reply:
Why do you think it would be a better name? The way I see it, the difference between `on_loop` and `_on_loop` is that `on_loop` does things that are common to all backends, as well as calling `_on_loop`, and `_on_loop` does things that are specific to the implementation of the backend. There's not really any difference in forcefulness between the two methods

Reply:
Oops, that comment was directed to the `on_loop` method. I suggested this because we only use the `on_loop` method in `overlay.py`. It is not used elsewhere. To me, it feels like we're just forcing an update hence the name.

Reply:
Hmm. I still can't say I agree with the name. `on_loop` doesn't just update the display, but it does miscellaneous stuff too, like sending recording statuses. Happy to change it though if you think it's better

Reply:
After thinking about this a bit more, I can see what you're getting at now. So happy to resolve this too!

Reply:
> Derived classes override `_on_loop` here but not `on_loop`, but in other places where there aren't two methods, derived classes would override a method without an underscore. Is this fine, or is there a convention that makes more sense I should use?

So based on my understanding is that the reason why you made these separate is in case a cam had different functionality. Someone from my FIT1008 class mentioned this useful function `super()` which allows you to use code from methods of the parent class that are under the same name. However, it'll require someone writing the child class to know that there is an existing parent class and need to use `super()`. 

More info here: https://stackoverflow.com/questions/37679421/adding-extra-functionality-to-parent-class-method-without-changing-its-name

> Hmm. I still can't say I agree with the name. `on_loop` doesn't just update the display, but it does miscellaneous stuff too, like sending recording statuses. Happy to change it though if you think it's better

Prefer keeping `on_loop` and `_on_loop` separate however. Makes it easier to understand that `on_loop` is consistent for all cams, while `_on_loop` can be modified for different cams. Although, I think they should be swapped, since one leading underscore means that it should be a private method (please check my understanding of this haha).

Reply:
> However, it'll require someone writing the child class to know that there is an existing parent class and need to use `super()`.

That's exactly why I wrote it this way around. If the client calls a method of superclass, they know that all the code will run. If a client calls something specific in the implementation classes, they need to just trust that it's been written correctly.

> Although, I think they should be swapped, since one leading underscore means that it should be a private method

Python is quite weird in its "private" methods. The closest thing it has is methods which start with a double underscore `__` which are impossible (without going out of your way) to call from outside the class. This would be great here, except this would mean that the superclass can't even call the child classes implementation. Instead, we use a single underscore, which has no meaning built into the language of python, but is generally accepted to indicate "don't call this method unless you really know what you're doing".

By the way, many big OO languages (C++, C#, Java...) have the concept of not just public and private methods, but also protected methods, which means the method can be called by super/sub classes, but not externally. Exactly what we'd want here, alas, it's not in Python.

The reason why I have the methods the way around that they are is that I expect clients accessing some backend to call `on_loop`, because this ensures that both the stuff in `on_loop` and `_on_loop` in the implementation class is completed. Use cases for only running the implementation-specific stuff in `_on_loop` would be very rare, if existent at all, which is why I prefixed it with the "don't call unless you know what you're doing" underscore.